### PR TITLE
SwiftRemoteMirror: Lower types with imported / Objective-C types

### PR DESCRIPTION
**Pull request by bitjammer**
- Lower Objective-C class typerefs as strong references with unknown
  reference counting.
- Lower other imported C types as builtin blobs of their known
  size, alignment, etc.

In the future, it might be beneficial to track which stored properties
of imported types are pointers, for better conservative scanning of
outgoing pointers to the heap.

rdar://problem/26240258
rdar://problem/26240394

### CODE REVIEW

**Comment by bitjammer** [comment]
@slavapestov Can you double-check me, please?

**Comment by bitjammer** [comment]
@swift-ci Please smoke test

File: include/swift/Reflection/Records.h [hunk 1]
```diff
@@ -110,6 +110,8 @@ enum class FieldDescriptorKind : uint16_t {
   Protocol,
   ClassProtocol,
   ObjCProtocol,
+  ObjCClass,
+  Imported
```
Comment (slavapestov) [inline include/swift/Reflection/Records.h]: Splitting this up into ImportedStruct and ImportedEnum might be useful later

**Comment by bitjammer** [comment]
@swift-ci Please smoke test

File: lib/IRGen/GenReflection.cpp [hunk 2]
```diff
@@ -345,6 +352,18 @@ class FieldTypeMetadataBuilder : public ReflectionMetadataBuilder {
     auto type = NTD->getDeclaredType()->getCanonicalType();
     addTypeRef(NTD->getModuleContext(), type);
 
+    if (NTD->hasClangNode()) {
+      auto CD = dyn_cast<ClassDecl>(NTD);
+      if (CD && CD->isObjC()) {
```
Comment (slavapestov) [inline lib/IRGen/GenReflection.cpp]: Might look better to do

    if (isa<ClassDecl>(NTD) && NTD->isObjC())

File: lib/IRGen/GenReflection.cpp [hunk 3]
```diff
@@ -790,11 +807,18 @@ void IRGenModule::emitBuiltinReflectionMetadata() {
   if (!IRGen.Opts.EnableReflectionBuiltins)
```
Comment (slavapestov) [inline lib/IRGen/GenReflection.cpp]: You don't need this flag at all anymore then -- the original purpose was to enable emitting metadata for the various "special" builtins that are not directly referenced but still needed.

File: include/swift/Reflection/Records.h [hunk 4]
```diff
@@ -110,6 +110,8 @@ enum class FieldDescriptorKind : uint16_t {
   Protocol,
   ClassProtocol,
   ObjCProtocol,
+  ObjCClass,
```
Comment (slavapestov) [inline include/swift/Reflection/Records.h]: Can you use ObjCClass for Swift-defined classes that inherit from NSObject, too? (I believe IRGen has utility functions to get the reference counting style without you walking the inheritance chain directly; just map Unknown to ObjCClass).

File: include/swift/Reflection/Records.h [hunk 5]
```diff
@@ -110,6 +110,8 @@ enum class FieldDescriptorKind : uint16_t {
   Protocol,
```
Comment (slavapestov) [inline include/swift/Reflection/Records.h]: Some comments on the members of this enum would be nice, explaining exactly how the rest of the field descriptor is interpreted in each case.

**Comment by bitjammer** [comment]
@swift-ci Please test

**Comment by bitjammer** [comment]
Got a fix for the Linux test. The OS X bot seems to be backed up, so I'm going to launch an all-platforms test run locally.

**Comment by bitjammer** [comment]
All Apple platform tests pass locally - checking Linux.

**Comment by bitjammer** [comment]
@swift-ci Please test Linux

**Pull request merged**